[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68149

JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32773666](https://our.internmc.facebook.com/intern/diff/D32773666)

davidberard98 committed:
[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp

JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

eellison commented:
deferring to @malfet and @desertfire

malfet commented on torch/csrc/jit/passes/frozen_conv_add_relu_fusion_impl.cpp:
```diff
@@ -0,0 +1,127 @@
+#include <torch/csrc/jit/passes/graph_rewrite_helper.h>
+#include <torch/csrc/jit/passes/remove_mutation.h>
+#include <torch/csrc/jit/passes/subgraph_rewrite.h>
+#ifdef USE_CUDA
```
You don't really need `USE_CUDA` here, as this file should not even be compiled for the CPU, shouldn't it?

davidberard98 replied:
In the internal build I can see why that is true - but how does that work in the OSS build?

malfet commented on torch/csrc/jit/passes/frozen_conv_add_relu_fusion_impl.cpp:
```diff
@@ -0,0 +1,127 @@
+template <typename T>
+struct ConvAddReluRegisterer {
+  ConvAddReluRegisterer(T& fn) {
+    fn = fuseFrozenConvAddReluImpl;
```
It should be other way around, shouldn't it?
```suggestion
    fuseFrozenConvAddReluImpl = fn;
```

davidberard98 replied:
_fuseFrozenConvAddReluImpl is the function<> object (passed in as T& fn) and fuseFrozenConvAddReluImpl is the actual implementation defined in frozen_.._impl.cpp. But it's a lot clearer with your suggestion.

malfet commented on torch/csrc/jit/passes/frozen_conv_add_relu_fusion_impl.cpp:
```diff
@@ -0,0 +1,127 @@
+#endif
+}
+
+template <typename T>
+struct ConvAddReluRegisterer {
+  ConvAddReluRegisterer(T& fn) {
+    fn = fuseFrozenConvAddReluImpl;
+  }
+};
+
+static ConvAddReluRegisterer<decltype(_fuseFrozenConvAddReluImpl)>
+    convAddReluRegisterer(_fuseFrozenConvAddReluImpl);
```
This would be much more compact, isn't it?
```suggestion
auto dummyInitializer = []() {
    fuseFrozenConvAddReluImpl = _fuseFrozenConvAddReluImpl;
    return true;
  }();
```

davidberard98 replied:
Yeah, that is a lot better - thanks!

malfet commented:
To clarify on requesting changes - global initializer, as it is currently implemented, is wrong, isn't it?
Otherwise, LGTM, though again, `#ifdef USE_CUDA` guards are not needed for GPU-only code (though perhaps it would be better to add name `cuda` to the file that implements the pass.

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

eellison reviewed (approved):
seems like comments are addressed

desertfire commented:
LGTM in general. One question is how extensible this solution is. Say if we want to add a similar fusion pass for MKLDNN (https://github.com/pytorch/pytorch/pull/64639), would the initializations (MKLDNN and CUDA) conflict with each other?

davidberard98 commented:
@desertfire afaik this could be extended with other passes (maybe use a vector<function<>> instead) without issues

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Separate GPU implementation of frozen_conv_add_relu_fusion.cpp"


JIT optimization passes are part of the CPU-only build (i.e. necessary GPU flags are not passed in). This separates the implementation of frozen_conv_add_relu_fusion so that the GPU-enabled implementation is registered at runtime (if it is available)

Differential Revision: [D32329330](https://our.internmc.facebook.com/intern/diff/D32329330/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32329330/)!

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32773666).

xsacha commented:
In a CPU-only build of Torch, I get these warnings on every run of a new model:
[W C:\work\pytorch\torch\csrc\jit\passes\frozen_conv_add_relu_fusion.cpp:29] Warning: No definition of _fuseFrozenConvAddReluImpl found (function FuseFrozenConvAddRelu)

Is this normal? If it is normal to hit this path, should it not be silent instead?

davidberard98 commented:
Removing the warning in #72441
